SharpPcap/WinDivert/WinDivertDevice.cs

```diff
@@ -112,6 +112,52 @@ public RawCapture GetNextPacket()
             }
         }
 
+        /// <summary>
+        /// Return 0 upon success, non-zero upon error
+        /// </summary>
+        /// <returns></returns>
+        private int SendPacketArrivalEvent()
+        {
+            ThrowIfNotOpen();
+            Span<byte> buffer = stackalloc byte[4096];
+            while (true)
+            {
+                bool ret;
+                WinDivertAddress addr;
+                uint readLen;
+                unsafe
+                {
+                    fixed (byte* p = buffer)
+                    {
+                        ret = WinDivertNative.WinDivertRecv(Handle, new IntPtr(p), (uint)buffer.Length, out readLen, out addr);
+                    }
+                }
+                if (!ret)
+                {
+                    var err = Marshal.GetLastWin32Error();
+                    if (err == ERROR_INSUFFICIENT_BUFFER)
+                    {
+                        // Increase buffer size
+                        buffer = stackalloc byte[buffer.Length * 2];
+                        continue;
+                    }
+                    if (err == ERROR_NO_DATA)
+                    {
+                        return err;
+                    }
+                    ThrowWin32Error("Recv failed", err);
+                }
+                var timestamp = new PosixTimeval(BootTime + TimeSpan.FromTicks(addr.Timestamp));
+                var data = buffer.Slice(0, (int)readLen).ToArray();
+
+                // TODO: how do we pass back WinDivert specific info?
```

Comment:
WinDivert uses `WinDivertCapture` to pass WinDivert specific info

Reply:
Since we aren't passing a derived class from RawCapture through the capture args thoughts on how to do that for WinDivert?

Reply:
Create a constructor for CaptureEventArgs that accepts a RawCapture

Reply:
Doesn't that break the performance gain from Span<> with WinDivert as we'd be copying data for the WinDivertCapture instance creation?

Reply:
Something like this may work then
```cs
    public ref struct CaptureEventArgs
    {
        private RawCapture rawCapture;
        public RawCapture Packet {
            get
            {
                return rawCapture ?? (rawCapture = new RawCapture(Device, Header, Data));
            }
        }
        public ICaptureDevice Device { get; }

        public ICaptureHeader Header { get; }

        public Span<byte> Data { get; }
    }
```
where `ICaptureHeader` is either a `PcapHeader` or a `WinDivertHeader` that holds the necessary metadata.

Reply:
@kayoub5 thoughts on collapsing PcapDevice.GetNextPacket() down into a single one like PcapDevice.GetNextPacket(out CaptureEventArgs)? We could do the same on the WinDivert side to keep from adding an extra method with the Span<> changes.

Reply:
It could be charged to an extension method

Reply:
See what we did for `SendPacket`, only one method in the Device class, and the rest implemented as extension methods in [CaptureDeviceExtensions](https://github.com/chmorgan/sharppcap/blob/master/SharpPcap/CaptureDeviceExtensions.cs#L57)